### PR TITLE
Fixed thread affinity more 64 cores before Windows 11 and Windows Server 2022

### PR DESCRIPTION
**Pull request by GermanAizek**
@mrexodia Hi,
This is solution to this problem for older Windows platforms https://github.com/x64dbg/x64dbg/pull/3272#issuecomment-1817674574

### CODE REVIEW

**Comment by GermanAizek** [comment]
> What exactly does this solve again?

https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2023-0/thread-affinity-interface.html
https://www.anandtech.com/show/15483/amd-threadripper-3990x-review/3
https://github.com/OpenHFT/Java-Thread-Affinity/issues/28

File: src/dbg/_global.cpp [hunk 1]
```diff
@@ -396,6 +396,14 @@ void WaitForMultipleThreadsTermination(const HANDLE* hThread, int count, DWORD t
         CloseHandle(hThread[i]);
 }
 
+void setThreadAffinityAllGroupCores(HANDLE handle)
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/dbg/_global.cpp:399]: > Over 99% of our users have less than 32 cores. Could we disable these for machines with too few cores to make the overhead smaller?

@torusrxxx yep, u correctly proposed idea to check count sockets from user or physical processors

**Comment by ZehMatt** [comment]
I have to agree with @mrexodia, x64dbg has a lot of locking in place and there is never a lot of work done in parallel and the few times it does its typically I/O bound, spawning as many threads as possible is not always ideal unless you really want to some crunch some data in parallel, in this case its a waste of resources. x64dbg would be pretty much the same if were to limit it to only 4 threads, so I'm against this rather pointless change, sorry.

**Comment by mrexodia** [comment]
Thanks, but I don't think the benefit of this change was sufficiently demonstrated.

**Comment by GermanAizek** [comment]
> I honestly don't quite see the issue here. Most of the threads in x64dbg are doing mostly trivial work (except perhaps the symbol loading thread, but this one is disk I/O bound). What's the difference if there will be 64 or 128 threads available to x64dbg? Do you have any performance measurements that are actually improved?
> 
> The problem for me is that these changes are quite invasive to the codebase, so it would be good to have a clear, measurable impact to motivate them. The articles you linked are just very vague and speak about benchmark loads, we are dealing with a real-world load here...

@mrexodia If you look at Microsoft documentation for SetThreadAffinityMask function, or in general about Windows OS, There is no term here using cores by process, there are logical cores. Mention cores here refers to count logical cores, and given work Windows task scheduler, one logical core is given for one dedicated program thread.

A long time ago, conventional consumer Threadripper processors or used old servers beyond 32 cores (64 threads).

**Comment by mrexodia** [comment]
Yeah I understand there is some subtly with the scheduling, but the point is still the same: what is the _measured_ benefit of changing this? Fundamentally x64dbg is a single-threaded program and there isn’t much to gain by allowing more threads in parallel…

**Comment by GermanAizek** [comment]
Is there really no milestone for multithreading work in progress?

**Comment by GermanAizek** [comment]
For example, comparing two functions or files, multithreading is recommended if possible, imagine as bad analyzing large crackme 300 MB.

**Comment by mrexodia** [comment]
No, x64dbg doesn’t have those features 🤷‍♂️ And if they will be implemented in a plugin the plugin can use this API to speed things up.

**Pull request closed without merging**